torch/optim/adam.py

```diff
@@ -104,6 +110,9 @@ def step(self, closure=None):
                 bias_correction2 = 1 - beta2 ** state['step']
                 step_size = group['lr'] * math.sqrt(bias_correction2) / bias_correction1
 
+                if group['weight_decay'] != 0 and not group['l2_reg']:
+                    p.data.add_(-group['weight_decay'], p.data)
```

Comment:
```suggestion
                    p.data.add_(-group['weight_decay'] * group['lr'], p.data)
```

In L2, the weight decay is stored inside the grad variable, which is later multiplied by the learning rate. This new formulation is not multiplying it by the learning rate (see Algorithms 1 and 2 in the paper).

Reply:
I will reply to this comment only since the rest are the same... yes in the l2 reg you are right the weight decay is in the grad and it gets multiplied by the learning rate... and this is the default behaviour when `l2_reg = True`. But  in this block we are doing what the paper suggests which is the green bit in Algorithm 2 of the paper which only happens when `l2_reg=False`. Hope that clears the confusion?

Reply:
Algorithm 2 has this term inside parenthesis, it _is_ being multiplied by the learning rate over there, not here... ;)

Reply:
ah you mean the `\eta_t` in Algorithm 2? Remember `\eta_t` is the learning rate schedular not the learning rate, the learning rate in the paper is `\alpha`... hope that helps